plasmapy/analysis/tests/test_fit_functions.py

```diff
@@ -0,0 +1,497 @@
+        for name in ("curve_fit", "func", "func_err", "root_solve", "__call__"):
+            assert hasattr(self.ff_class, name)
+
+        # required properties
+        for name in (
+            "curve_fit_results",
+            "latex_str",
+            "params",
+            "param_errors",
+            "param_names",
+            "rsq",
+        ):
+            assert hasattr(self.ff_class, name)
+            assert isinstance(getattr(self.ff_class, name), property)
+
+    def test_abstractmethods(self):
+        # abstract methods
+        for name in ("__str__", "func", "func_err", "latex_str"):
+            assert name in self.ff_class.__abstractmethods__
+
+
+class BaseFFTests(ABC):
+    abc = ffuncs.AbstractFitFunction
+    _test_params = NotImplemented  # type: tuple
+    _test_param_errors = NotImplemented  # type: tuple
+
+    @property
+    @abstractmethod
+    def ff_class(self):
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def func(x, *args):
+        raise NotImplementedError
+
+    @abstractmethod
+    def func_err(self, x, params, param_errors, x_err=None):
+        raise NotImplementedError
+
+    def test_inheritance(self):
+        assert issubclass(self.ff_class, self.abc)
+
+    def test_iscallable(self):
+        assert callable(self.ff_class())
+
+    def test_basics(self):
+        assert hasattr(self.ff_class, "_param_names")
+        if self.ff_class._param_names == NotImplemented:
+            pytest.fail(
+                f"{self.ff_class} class attribute '_param_names' needs to "
+                f" be defined as a tuple of strings representing the names of "
+                f"the fit parameters."
+            )
+
+        # required attributes/methods
+        for name in ("curve_fit", "func", "func_err", "root_solve"):
+            assert hasattr(self.ff_class, name)
+
+        # required properties
+        for name in (
+            "curve_fit_results",
+            "latex_str",
+            "params",
+            "param_errors",
+            "param_names",
+            "rsq",
+        ):
+            assert hasattr(self.ff_class, name)
+            assert isinstance(getattr(self.ff_class, name), property)
+
+        foo = self.ff_class()
+        assert foo.__repr__() == f"{foo.__str__()} {foo.__class__}"
+
+    def test_instantiation(self):
+        # default
+        foo = self.ff_class()
+
+        assert isinstance(foo.param_names, tuple)
+        assert len(foo.param_names) != 0
+        assert all(isinstance(val, str) for val in foo.param_names)
+
+        assert hasattr(foo, "FitParamTuple")
+        assert issubclass(foo.FitParamTuple, tuple)
+        for name in foo.param_names:
+            assert hasattr(foo.FitParamTuple, name)
+
+        assert foo.curve_fit_results is None
+        assert foo.params is None
+        assert foo.param_errors is None
+        assert foo.rsq is None
+
+        assert isinstance(foo.latex_str, str)
```

Comment:
One possibility here would be to break this up so that there is one `assert` statement per test, so that any test failures a bit further down would still show up if one of the higher up tests fails.  For this case, I don't think that this would make a huge difference, so I'll leave it up to you.